Beam/Include/Beam/Threading/Sync.hpp

```diff
@@ -52,38 +130,53 @@ namespace Threading {
       */
       Sync(Sync&& sync);
 
-      //! Returns a reference of the value.
-      T& AcquireReference();
-
       //! Returns a copy of the value.
       T Acquire() const;
 
+      //! Acquire Sync's value in a synchronized manner.
+      /*!
+        \param f The action to perform on the value.
+      */
       template<typename F>
-      decltype(auto) With(F f) {
-        boost::unique_lock<Mutex> lock(m_mutex);
-        m_lock = &lock;
-        return f(m_value);
+      decltype(auto) With(const F& f) {
+        if constexpr(boost::hof::is_invocable<F, Value>()) {
+          return static_cast<const Sync*>(this)->With(f);
+        } else {
+          auto lock = WriteLock(m_mutex);
+          m_lock = LockProxy(&lock);
+          return f(m_value);
+        }
       }
 
+      //! Acquires Sync's value in a synchronized manner.
+      /*!
+        \param f The action to perform on the value.
+      */
       template<typename F>
-      decltype(auto) With(F f) const {
-        boost::unique_lock<Mutex> lock(m_mutex);
-        m_lock = &lock;
+      decltype(auto) With(const F& f) const {
+        auto lock = ReadLock(m_mutex);
+        m_lock = LockProxy(&lock);
         return f(m_value);
       }
 
+      //! Atomically acquires Sync's value along with another Sync's value.
+      /*!
+        \param s2 The other Sync.
+        \param f The action to perform on the values.
+      */
       template<typename S2, typename M2, typename F>
-      decltype(auto) With(Sync<S2, M2>& s2, F f) {
-        boost::unique_lock<Mutex> lock1(m_mutex, boost::defer_lock);
-        boost::unique_lock<M2> lock2(s2.m_mutex, boost::defer_lock);
+      decltype(auto) With(Sync<S2, M2>& s2, const F& f) {
+        auto lock1 = WriteLock(m_mutex, boost::defer_lock);
+        auto lock2 = typename Sync<S2, M2>::WriteLock(s2.m_mutex,
+          boost::defer_lock);
         boost::lock(lock1, lock2);
-        m_lock = &lock1;
-        s2.m_lock = &lock2;
+        m_lock = LockProxy(&lock1);
+        s2.m_lock = typename Sync<S2, M2>::LockProxy(&lock2);
         return f(m_value, s2.m_value);
       }
 
       //! Returns the lock used for synchronization.
-      boost::unique_lock<Mutex>& GetLock() const;
+      const LockProxy& GetLock() const;
```

Comment:
The returned lock can not be a const-ref, condition variables must be able to unlock it.